### ISSUE TITLE
Concurrency error when updating a settings part for content type

### ISSUE THREAD

**Issue by giannik** [OrchardCMS/OrchardCore#7449]
From the admin  a try to update/save the settings for a custom part In my content type.
When hitting Save button in the admin I get the following error  : 

![image](https://user-images.githubusercontent.com/1516981/97732179-d398cb00-1ade-11eb-8235-a6372377ed90.png)

Seems that it has to do with the recent Distributed cache pr  (#5249) that was merged

This is the code in my settingsdisplay driver : 
The updateAsync calls some  contentItemDisplayManager.UpdateEditorAsync() 
which i beleive is the culprit.

If i comment out this call then the error disappears.

```
 public class ActionsPanelPartSettingsDisplayDriver : ContentTypePartDefinitionDisplayDriver
    {
        private readonly IContentManager _contentManager;
        private readonly IServiceProvider _serviceProvider;
  
        public override async Task<IDisplayResult> UpdateAsync(ContentTypePartDefinition contentTypePartDefinition, UpdateTypePartEditorContext context)
        {
            if (!String.Equals(nameof(ActionsPanelPart), contentTypePartDefinition.PartDefinition.Name, StringComparison.Ordinal))
            {
                return null;
            }

            var contentItemDisplayManager = _serviceProvider.GetRequiredService<IContentItemDisplayManager>();

             var model = new ActionsPanelPartSettingsViewModel();
             await context.Updater.TryUpdateModelAsync(model, Prefix );
 

             for (var i = 0; i < model.Prefixes.Length; i++)
             {
                 var contentItem = await _contentManager.NewAsync(model.ContentTypes[i]);

                 contentItem.Weld(new FlowMetadata());

                var widgetModel = await    contentItemDisplayManager.UpdateEditorAsync(contentItem, context.Updater, context.IsNew, htmlFieldPrefix: model.Prefixes[i]);

                 model.FlowPartDisplayMode.Widgets.Add(contentItem);
            }
         
            context.Builder.WithSettings(new ActionsPanelPartSettings
                { FlowPartDisplayMode = model.FlowPartDisplayMode
               });

            return Edit(contentTypePartDefinition, context.Updater);
        }
    }
```

Any ideas @jtkech ?

**Comment by jtkech** [comment]
Yes there is a new `checkConcurrency` option that we now use when doing a YesSQL session save and that add an optimistic concurrency checking when commiting the session, this by using a new `Version` column in a document table.

This may happen when we update the same document (same `Version` value) concurrently

Did you see the same when updating another type part settings?

**Comment by giannik** [comment]
No , other part settings do not cause the issue.
Only this one which has a call to  : 
```
await    contentItemDisplayManager.UpdateEditorAsync()
```
Is there any workaround?
Thanks.

**Comment by giannik** [comment]
But also noticed in latest dev when enabling /disabling features from the admin i get the following, might be related : 

I am using sql server express as db

![image](https://user-images.githubusercontent.com/1516981/97775828-62075e00-1b6c-11eb-89df-858928343e03.png)

**Comment by jtkech** [comment]
Sorry i was not able to repro

Can you share a testing app so that i could repro

**About the MARS issue most of the time it is due to a non awaited async call in a driver / handler / index provider**

You are storing content type definitions in the database, not in the `ContentDefinition.json` file, right?

It is uncommon that a type part settings hold fake content items ;) Creating them use the content type definitions that you are updating in your type part definition driver, hmm, but it doesn't explain the issue.

**Comment by giannik** [comment]
Found out the culprit with the MARS issue. It was a non awaited call in one of my modules, as you said.
Closing the issue, eventually will figure out the type part settings and will report my findings.
Thank you @jtkech

**Comment by deanmarcussen** [comment]
@jtkech reopening this as am getting similar MARS related issues.

Here's a repro
- sql server (local)
- blog recipe
- export content definitions
- disable file content definition store
- import content definitions
- enable module with multiple migrations 

MARS error. 
When migration is tweaked a bit can turn into an open data reader exception (so it is like we had last time with missing awaits.

Here's a sample migration I can get to fail 3/4 times
```
        public int Create()
        {
            _contentDefinitionManager.AlterPartDefinition("SeoMetaPart", builder => builder
                .Attachable()
                .WithDescription("Provides a part that allows seo meta descriptions to be applied to a content item.")
                .WithField("DefaultSocialImage", field => field
                    .OfType("MediaField")
                    .WithDisplayName("Default social image")
                    .WithSettings(new MediaFieldSettings { Multiple = false }))
                .WithField("OpenGraphImage", field => field
                    .OfType("MediaField")
                    .WithDisplayName("Open graph image")
                    .WithSettings(new MediaFieldSettings { Multiple = false }))
                .WithField("TwitterImage", field => field
                    .OfType("MediaField")
                    .WithDisplayName("Twitter image")
                    .WithSettings(new MediaFieldSettings { Multiple = false }))
            );

            return 1;
        }

        public async Task<int> UpdateFrom1Async()
        {
            await _recipeMigrator.ExecuteAsync("socialmetasettings.recipe.json", this);
            return 2;
        }
```

[socialmetasettings.recipe.json.zip](https://github.com/OrchardCMS/OrchardCore/files/5489011/socialmetasettings.recipe.json.zip)

if it doesn't fail this helps make it fail
```
        public async Task<int> UpdateFrom1Async()
        {
            await Task.Delay(1000);
            await _recipeMigrator.ExecuteAsync("socialmetasettings.recipe.json", this);
            await Task.Delay(1000);
            return 2;
        }
```

Sometimes it seems like it needs to be delayed (I saw this first on a remote server connection to azure, which is always slower)

Adapting the migration to use
`CreateAsync` and the `AlterPartDefinitionAsync` methods was more likely to turn it into an open data reader exception.

Buried underneath the open data reader exception (was only visible in the logs), is the `ConcurrencyException`

Ping me if I can help with anymore details

**Comment by jptissot** [comment]
I ran into this as well this morning. 

Repro steps:

- Create a module with a migration that adds a role via a recipe migration and content types via the `ContentDefinitionManager`.
- If the recipe is executed before the `ContentDefinitionManager`, the exception is shown. If the recipe is executed after, no exception occur. 

This also happens with feature dependencies.
- Feature 1 creates a migration that sets a role via the `RoleManager<IRole>` or a recipe step
- Feature 2 uses the `ContentDefinitionManager`

Exception shown:
```
ConcurrencyException: The document could not be updated as it has been changed by another process.
YesSql.Commands.UpdateDocumentCommand.ExecuteAsync(DbConnection connection, DbTransaction transaction, ISqlDialect dialect, ILogger logger)

DocumentStoreCommitException: The 'ContentDefinitionRecord' could not be persisted and cached as it has been changed by another process.
```

**Comment by jtkech** [comment]
Yes, maybe different issues here but not so easy to repro, may be good to have a testing app to be able to repro each individual issue, and / or don't hesitate to ping me on skipe.

But "Good" new, I just could reproduce the `MultipleActiveResultSets` issue by adding a role through a migration recipe, maybe because roles are managed differently because of the related aspnetcore services, so at some point before doing a mutation we still first use a `GetOrCreateImmutableAsync()` that does a session `Detach()` on the entity to get an isolated object, so maybe in that case the YesSql Version is double incremented

Note: `ConcurrencyException` may also be thrown if the YesSql cmd fails to update (`updateCount` is not equal to 1).

**Update**: So yes if i remove the following

            //if (document != null)
            //{
            //    _session.Detach(document);
            //    return (true, document);
            //}

I don't get the `MultipleActiveResultSets` issue, but i now get the following (with database content definition)

    The 'ContentDefinitionRecord' could not be persisted and cached as it has been changed by another process.

So okay, i'm now able to repro both issues, i'm working on ;)

For now, about the `ConcurrencyException` there is a workaround, the `CheckConcurrency` option is true by default but esay to set it to false, idem for the `CheckConsistency` option that redo a `GetOrCreateImmutableAsync()` at the end of an update and that may explain a different issue (but maybe related to the same unique issue). This can be changed in `DocumentOptionsSetup.cs`, note: can be also changed only for a given type through the appsettings with a section named with the document type fullName.

**Comment by jtkech** [comment]
@jptissot @deanmarcussen

Just did a PR #7516, not sure at 100% so if you can give it a try to see if it fixes your issues

**Comment by giannik** [comment]
Tried the pr with my case  , using in part type settings editor : 
```
await    contentItemDisplayManager.UpdateEditorAsync()
```
but did not resolve issue.

**Comment by jtkech** [comment]
@giannik 

Yes, i think it is another use case as `NewAsync()` is getting definitions that you are updating, so let's @jptissot and @deanmarcussen try their use cases first, then tomorrow i will try to understand what happens in your case.

**Comment by deanmarcussen** [comment]
@jtkech tried with my scenario.

Without your branch fails.
With your branch works :)

**Comment by jtkech** [comment]
@deanmarcussen good to know!

@jptissot can you try it to confirm when you will have time

@giannik i tried the same kind of loop using `UpdateEditorAsync()` in our `FlowPartSettingsDisplayDriver`, with sql server and type definitions in the database, i can't repro, it works even without my change in the related PR

So it seems that something else happens, maybe in a custom handler, I could help if you have a testing app to repro, in the meantime you could try this workaround

        services.Configure<DocumentOptions>(typeof(ContentDefinitionRecord).FullName, options =>
        {
            options.CheckConcurrency = false;
        });

**Comment by giannik** [comment]
thanks @jtkech Ill try to make a testing app in the weekend and send it here.

**Comment by jtkech** [comment]
@giannik cool, thanks

@jptissot can you try #7516 if you have time, it works for @deanmarcussen, would be good to confirm it also fixes your issue ;)

**Comment by giannik** [comment]
@jtkech The error occurs from the workflows module here : 
```
namespace OrchardCore.Contents.Workflows.Handlers
{
 public class ContentsHandler : ContentHandlerBase
    {
            public override Task UpdatedAsync(UpdateContentContext context)
         {
             return TriggerWorkflowEventAsync(nameof(ContentUpdatedEvent), context.ContentItem);
          }

    }
}
```

and specifically from this line when return is called : 

 https://github.com/OrchardCMS/OrchardCore/blob/7a2e676b44dd1c9070c9b8495e6a6c07aa99c941/src/OrchardCore.Modules/OrchardCore.Workflows/Services/WorkflowManager.cs#L124

**Comment by jtkech** [comment]
@jptissot can you try #7516 if you have time, it works for @deanmarcussen, would be good to confirm it also fixes your issue ;)

@giannik yes the `NewAsync()` and `UpdateEditorAsync()` in your loop in your `ContentTypePartDefinitionDisplayDriver` will trigger workflow events, i will think about this use case this night.

**Do you have any workflow using a `ContentUpdatedEvent`**?

**Comment by giannik** [comment]
No i dont have any code using  *ContentUpdatedEvent. *
*As seen in the return call which gets called when **no workflow matches
the event.*

**Comment by jtkech** [comment]
@giannik okay thanks, i assume this is why i was not able to repro as i didn't enable the `WorkFlows` feature, hmm maybe i've some ideas, i will re-try it this night.

**Comment by jtkech** [comment]
@giannik "Good" new, just tried, now i can repro ;)

What i can see right now, is that on each loop `TriggerWorkflowEventAsync()` really does 2 database queries, this is not the case when we get multiple times a content item or a document as type definitions where we return an already loaded object.

**Comment by jtkech** [comment]
@giannik okay i think i found a little issue on YesSql side

There is a case where the document version is incremented but the related `UpdateDocumentCommand` is not enlisted, i will open an issue and maybe a PR on Yesql.

Meanwhile, for your use case you can use the following workaround in the `OC.ContentTypes` module startup

    services.Configure<DocumentOptions>(typeof(ContentDefinitionRecord).FullName, options =>
    {
        options.CheckConcurrency = false;
    });

If you don't use the full source code, i can show you how to do it through the appsettings.json, let me know

**Update**: https://github.com/sebastienros/yessql/pull/287

**Comment by giannik** [comment]
@jtkech thank you very very much.
I dont mind wating for the yessql update.
Must buy you a beer.

**Comment by giannik** [comment]
also when trying to add a new tenant from the admin and setting it up i get the following error : 
Also applied the pull #7516   but still the same. Think it has to do with @jptissot case.


![image](https://user-images.githubusercontent.com/1516981/98449649-d11d1d80-213d-11eb-81fc-223f0e1e21f8.png)

**Comment by jtkech** [comment]
@giannik 

It doesn't find `tbl_ContentItemIndex`, not related to a single shared `Document`

So better to open another issue. 

Sorry I can't repro. Which setup recipe? Does it happen every time you create a tenant or only once? 

---

So, maybe i will merge #7516 as it fixes the issue of @deanmarcussen 

For your original issue it will be fixed through sebastienros/yessql#287

Merging #7516 will close this issue but feel free to open any new issue e.g. related to your last problem

**Comment by giannik** [comment]
sure , will open it as a new issue.

**Comment by jtkech** [comment]
@giannik 

Okay i found a temporary fix for your original issue (not the last different one), i integrated it in #7516 so that we don't have to wait for an update of YsSql

**Comment by jptissot** [comment]
@jtkech I just tested this today and it fixes the issue I was having.

**Comment by jtkech** [comment]
@jptissot cool, thanks for testing ;)
